File: CS/WinSolution.Module/DomainObject1.cs
using System;

using DevExpress.Xpo;

using DevExpress.ExpressApp;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Persistent.Validation;
using System.ComponentModel;
using DevExpress.Data.Filtering;

namespace WinSolution.Module {
    [DefaultClassOptions]
    public class DomainObject1 : BaseObject {
        public DomainObject1(Session session) : base(session) { }
        private string _name;
        public string Name {
            get { return _name; }
            set { SetPropertyValue("Name", ref _name, value); }
        }
        private MyRole _myRole;
        [ImmediatePostData]
        public MyRole MyRole {
            get { return _myRole; }
            set { SetPropertyValue("MyRole", ref _myRole, value); }
        }
        private MyUser _myUser;
        [DataSourceProperty("MyUserDataSource")]
        //[DataSourceCriteria("MyRoles[RoleName = 'r2']")]
        public MyUser MyUser {
            get { return _myUser; }
            set { SetPropertyValue("MyUser", ref _myUser, value); }
        }
        private XPCollection<MyUser> myUserDataSourceInternal = null;
        [Browsable(false), MemberDesignTimeVisibility(false)]
        public XPCollection<MyUser> MyUserDataSource {
            get {
                if (myUserDataSourceInternal == null) {
                    myUserDataSourceInternal = new XPCollection<MyUser>(Session);
                }
                UpdateAddressDataSource();
                return myUserDataSourceInternal;
            }
        }
        private void UpdateAddressDataSource() {
            myUserDataSourceInternal.Criteria = MyRole != null ? CriteriaOperator.Parse("MyRoles[RoleName = ?]", MyRole.RoleName) : null;
        }
    }
}

File: VB/WinSolution.Module/DomainObject1.vb
Imports Microsoft.VisualBasic
Imports System

Imports DevExpress.Xpo

Imports DevExpress.ExpressApp
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Persistent.Validation
Imports System.ComponentModel
Imports DevExpress.Data.Filtering

Namespace WinSolution.Module
	<DefaultClassOptions> _
	Public Class DomainObject1
		Inherits BaseObject
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Private _name As String
		Public Property Name() As String
			Get
				Return _name
			End Get
			Set(ByVal value As String)
				SetPropertyValue("Name", _name, value)
			End Set
		End Property
		Private _myRole As MyRole
		<ImmediatePostData> _
		Public Property MyRole() As MyRole
			Get
				Return _myRole
			End Get
			Set(ByVal value As MyRole)
				SetPropertyValue("MyRole", _myRole, value)
			End Set
		End Property
		Private _myUser As MyUser
		'[DataSourceCriteria("MyRoles[RoleName = 'r2']")]
		<DataSourceProperty("MyUserDataSource")> _
		Public Property MyUser() As MyUser
			Get
				Return _myUser
			End Get
			Set(ByVal value As MyUser)
				SetPropertyValue("MyUser", _myUser, value)
			End Set
		End Property
		Private myUserDataSourceInternal As XPCollection(Of MyUser) = Nothing
		<Browsable(False), MemberDesignTimeVisibility(False)> _
		Public ReadOnly Property MyUserDataSource() As XPCollection(Of MyUser)
			Get
				If myUserDataSourceInternal Is Nothing Then
					myUserDataSourceInternal = New XPCollection(Of MyUser)(Session)
				End If
				UpdateAddressDataSource()
				Return myUserDataSourceInternal
			End Get
		End Property
		Private Sub UpdateAddressDataSource()
			If Not MyRole Is Nothing Then
				myUserDataSourceInternal.Criteria = CriteriaOperator.Parse("MyRoles[RoleName = ?]", MyRole.RoleName)
			Else
				myUserDataSourceInternal.Criteria = Nothing
			End If
		End Sub
	End Class
End Namespace
